pkg/apis/kudo/v1alpha1/frameworkversion_types.go

```diff
@@ -61,9 +61,9 @@ const Parallel Ordering = "parallel"
 
 //Plan specifies a series of Phases that need to be completed.
 type Plan struct {
-	Strategy Ordering `json:"strategy"`
+	Strategy Ordering `json:"strategy" validate:"required,gt=0"` // makes field mandatory and checks if set and non empty
 	//Phases maps a phase name to a Phase object
-	Phases []Phase `json:"phases"`
+	Phases []Phase `json:"phases" validate:"gt=0,dive"` // makes field mandatory and checks if its gt 0
```

Comment:
Is this dive needed?

Reply:
Should that field be `required`?

Reply:
> Is this dive needed? It feels weird at the end of the chain?

`Phases []Phase 'json:"phases" validate:"gt=0,dive"'`

[GreaterThan](https://godoc.org/gopkg.in/bluesuncorp/validator.v8#hdr-Greater_Than) looks at the slice and ensures it has at least one item - it doesn't check if its an empty item or anything else with the item is wrong. Therefor we have `dive`.

[Dive](https://godoc.org/gopkg.in/go-playground/validator.v9#hdr-Dive) is used here to ensure that the values of the `Phase` struct will be validated after its validate tags. If I leave `dive` out here what happens it just looks at the slice, but I need it one level deeper. The test would fail as each `Phase`item in the slice wouldn't be validated anymore:
```
=== RUN   TestFrameworkType_Validation
--- FAIL: TestFrameworkType_Validation (0.00s)
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[0].Name:required: cannot be empty
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[0].Strategy:required: cannot be 
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[0].Steps:gt: cannot be []
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[1].Steps[0].Name:required: cannot be empty
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[1].Steps[0].Tasks:required: cannot be []
    frameworkversion_types_test.go:408: Missed expected error: ServiceSpec.Plans[plan2].Phases[1].Steps[1].Tasks[0]:gt: cannot be empty
FAIL
```

You don't find at the end of the chain `dives` in the official examples because they deal already with the final string data type. E.g.
```
[][]string with validation tag "gt=0,dive,len=1,dive,required"
// gt=0 will be applied to []
// len=1 will be applied to []string
// required will be applied to string
```
There is no nested level in `string` so that is why after `dive` a `required` is applied to the string type. Without the `dive` tag `required` would just be applied to `[]string`. While it says for [validate.Struct](https://github.com/go-playground/validator/blob/v8.18.2/validator.go#L498) that it will automatically validate nested structs I noticed the opposite. I filed already an issue with https://github.com/go-playground/validator/issues/458 but for the desired outcome it requires us right now to use `dive` at the end to kick off the validation of a nested struct imo.